scipy/_lib/_array_api.py

```diff
@@ -0,0 +1,135 @@
+
+
+# SCIPY_ARRAY_API, array_api_dispatch is used by sklearn
+array_api_dispatch = os.environ.get("array_api_dispatch", False)
+SCIPY_ARRAY_API = os.environ.get("SCIPY_ARRAY_API", array_api_dispatch)
+
+_GLOBAL_CONFIG = {"SCIPY_ARRAY_API": SCIPY_ARRAY_API}
+
+
+def compliance_scipy(*arrays):
+    for array in arrays:
+        if isinstance(array, np.ma.MaskedArray):
+            raise TypeError("'numpy.ma.MaskedArray' are not supported")
+        elif isinstance(array, np.matrix):
+            raise TypeError("'numpy.matrix' are not supported")
+
+
+def array_namespace(*arrays, single_namespace=True):
+    compliance_scipy(*arrays)
+
+    if not _GLOBAL_CONFIG["SCIPY_ARRAY_API"]:
+        return np
+
+    # if we cannot get the namespace, np is used
+    # here until moved upstream
+    namespaces = set()
+    for array in arrays:
+        try:
+            namespaces.add(array_api_compat.array_namespace(array))
+        except TypeError:
+            namespaces.add(array_api_compat.numpy)
+
+    if single_namespace and len(namespaces) != 1:
+        raise ValueError(
+            f"Expected a single common namespace for array inputs, \
+              but got: {[n.__name__ for n in namespaces]}"
+        )
+
+    (xp,) = namespaces
+
+    return xp
+
+
+def asarray(array, dtype=None, order=None, copy=None, *, xp=None):
+    """Drop-in replacement for `np.asarray`.
+
+    Memory layout parameter `order` is not exposed in the Array API standard.
+    `order` is only enforced if the input array implementation
+    is NumPy based, otherwise `order` is just silently ignored.
+
+    The purpose of this helper is to make it possible to share code for data
+    container validation without memory copies for both downstream use cases
+    """
+    if xp is None:
+        xp = array_namespace(array)
+    if xp.__name__ in {"numpy", "array_api_compat.numpy", "numpy.array_api"}:
+        # Use NumPy API to support order
+        if copy is True:
+            array = np.array(array, order=order, dtype=dtype)
+        else:
+            array = np.asarray(array, order=order, dtype=dtype)
+
+        # At this point array is a NumPy ndarray. We convert it to an array
+        # container that is consistent with the input's namespace.
+        return xp.asarray(array)
+    else:
+        return xp.asarray(array, dtype=dtype, copy=copy)
+
+
+def asarray_namespace(*arrays):
+    """Validate and convert arrays to a common namespace.
+
+    Parameters
+    ----------
+    *arrays : sequence of array_like
+        Arrays to validate and convert.
+
+    Returns
+    -------
+    *arrays : sequence of array_like
+        Validated and converted arrays to the common namespace.
+    namespace : module
+        Common namespace.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> x, y, xp = asarray_namespace([0, 1, 2], np.arange(3))
+    >>> xp.__name__
+    'array_api_compat.numpy'
+    >>> x, y
+    (array([0, 1, 2]]), array([0, 1, 2]))
+
+    """
```

Comment:
This function and the structure looks pretty good to me. Maybe a few comments here would help, to make it easier to understand what happens in what order. I think it is:
- this function, `asarray_namespace`, is the function meant to be called from each public function in a SciPy submodule
- `array_namespace` will: 
  - first use `compliance_scipy` to raise exceptions on known-bad subclasses
  - check for the global switch (_note that this is in the wrong order, `compliance_scipy` should be behind the global switch_).
  - then determine the namespace, without doing any coercion of array(-like) inputs
- the for-loop will call `xp.asarray`. _Comment: it would be cleaner to call `xp.asarray` directly, rather than `asarray`. All the `order`/`dtype`/`copy` handling there does nothing here._

Reply:
Ok I will add more comments.

Good point for the order, I will document that this would mean: one could use this on a code which supports "non-compliant" arrays when the flag is disabled.

> the for-loop will call xp.asarray. Comment: it would be cleaner to call xp.asarray directly, rather than asarray. All the order/dtype/copy handling there does nothing here.

Right 👍 and if one wants to get more control, they have `asarray` to call manually.

For `arrays = list(arrays)  # probably not good` I am not 100 sure but this seems to be ok:

```python
x, y = [0, 1, 2], np.arange(3)
id(x), id(y)
Out[]: (5047891008, 5059379632)
arrays = (x, y)
id(arrays[0]), id(arrays[1])
Out[]: (5047891008, 5059379632)
arrays = list(arrays)
id(arrays[0]), id(arrays[1])
Out[]: (5047891008, 5059379632)
```